core/src/main/java/io/grpc/Metadata.java

```diff
@@ -283,8 +395,25 @@ public boolean containsKey(Key<?> key) {
    */
   @ExperimentalApi
   public <T> void discardAll(Key<T> key) {
```

Comment:
@carl-mastrangelo 

I was suggesting to have something like that. Wouldn't that work?
```java
public <T> void discardAll(Key<T> key) {
  removeAll0(key, false);
}

public <T> Iterable<T> removeAll(Key<T> key) {
  return removeAll0(key, true);
}

private <T> Iterable<T> removeAll0(Key<T> key, boolean returnIterable) {
  ...
  if (returnIterable) {
    return new ValuableIterable(..)
  }
  return null;
}
```

**Offtopic:**
> The JVM is not smart enough to skip creating the Iterable when it isn't used by the caller of removeAll. (which is why ti showed up in the profile). 

I didn't look in this particular case, but I believe YourKit is known to mess with escape analysis. See for example http://psy-lob-saw.blogspot.de/2014/12/the-escape-of-arraylistiterator.html

Also previously `removeAll` was backed by `LinkedHashMap`. Looking with javap `HasMap.remove(...)` has a bytecode size of 315 (25 + 290), a few bytes codes added by `Metadata.removeAll()` and it comes close to the max inlining size of 325. So it might well be that `removeAll` couldn't be inlined before and thus the allocation could not be removed.

I wrote this little benchmark, that just allocates a new iterator. Once with inlining (`FooBenchmark.no_alloc`) and once with inlining disabled (`FooBenchmark.no_alloc_but_alloc`). I used JMH GC profiler to show that when the method is inlined the allocation is removed, else it's not.

https://gist.github.com/buchgr/9e5504b314ff302841f4b536de8e629b